### DISCUSSION TITLE
What is the best way to create components types?inheritance between types?

### DISCUSSION THREAD

**Discussion by jcenlo** [GrapesJS/grapesjs#4688] [Q&A] [answered]
Hi i am trying to create some components based in the bootstrap 4 plugin , i am rewriting to adapt to bootstrap 5 and new version of Grapesjs.

NOTE: the code works like is now, but i would like to know if is correct

Also i am trying to create "traits" components to extend other components like for example

TYPE  => TextDefault, i am creating this type to extend other components like ( headers, paragraphs ...) 
in the textDefault component  i am creating traits like (Text color, Font Family, Font Size...) everything align with the Fonts controller, i know i have the StyleManager to solve this but, in this traits i can configure limited options to follow bootstrap framework classes and styles.

THIS IS MY DEFAULT TEXT COMPONENT
This type will be "extend" in the header & paragraph to add this traits to the other component
`
import contexts from "../../bootstrap-contexts";
import { capitalize } from "../../utils";

export default (domc) => {
    const contexts_w_white = contexts.concat(["white"]);
    const textType = domc.getType("text");

    domc.addType("textDefault", {

        extend: "text",
        // Model definition
        model: {
            // Default properties
            defaults: {
                traits: [
                    // Text Color--------------------------------------------------
                    {
                        type: "class_select",
                        options: [
                            { value: "", name: "Default" },
                            ...contexts_w_white.map(function (v) {
                                return {
                                    value: "text-" + v,
                                    name: capitalize(v),
                                };
                            }),
                        ],
                        label: "Text color",
                    },

                    // Font Family --------------------------------------------------
                    {
                        type: "class_select",
                        options: [
                            { value: "", name: "Default" },
                            { value: "ff-primary", name: "fontFamily 1" },
                            { value: "ff-secondary", name: "fontFamily 2" },
                        ],
                        label: "Font Family",
                        name: "fontFamily",
                        changeProp: 1,
                    },
                    // Font Size --------------------------------------------------
                    {
                        type: "class_select",
                        options: [
                            { value: "", name: "None" },
                            { value: "fs-1", name: "1" },
                            { value: "fs-2", name: "2" },
                            { value: "fs-3", name: "3" },
                            { value: "fs-4", name: "4" },
                            { value: "fs-5", name: "5" },
                            { value: "fs-6", name: "6" },
                            { value: "small", name: "Small" },
                            { value: "lead", name: "Lead" },
                        ],
                        label: "Font Size",
                    },
                ].concat(textType.model.prototype.defaults.traits),
            },
        },
    });
};
`

THIS IS MY PARAGRAPH COMPONENT
In this component i am adding the traits of the 'textDefault' type and extend the 'text' type for the view
`import paragraphIcon from "raw-loader!../../icons/paragraph-solid.svg";

export const ParagraphBlock = (bm, label) => {
    bm.add("paragraph", {
        label: `
            ${paragraphIcon}
            <div>${label}</div>
        `,
        category: "Typography",
        content: {
            type: "paragraph",
            content:
                "Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor. Duis mollis, est non commodo luctus.",
        },
    });
};
export default (domc) => {
    const textDefaultType = domc.getType("textDefault");
    const textType = domc.getType("text");
    domc.addType("paragraph", {

        isComponent(el) {
            if (el && el.tagName && el.tagName === "P") {
                return { type: "paragraph" };
            }
        },

        model: {
            defaults: {
                "custom-name": "Paragraph",
                tagName: "p",
                traits: [
                    {
                        type: "class_select",
                        options: [
                            { value: "", name: "No" },
                            { value: "lead", name: "Yes" },
                        ],
                        label: "Lead?",
                    },
                // extend traits from textDefault
                ].concat(textDefaultType.model.prototype.defaults.traits),
            },
        },
        view: textType.view,
    });
};
`
I have a few questions:
1.- is this the best way to do this?
2.- i am not sure 100% how to use the extend function and what exactly extends
3.- in the trait JSON i am concat the textDefaultTypeTraits,¿what do you think about this?, is the correct way to do it

**Comment by artf** [comment, marked as answer]
Hi @jcencina I guess the best way to understand what is happening with the `extend` is to check the [code](https://github.com/artf/grapesjs/blob/01352d366dbbd846e8dd0d29249062814cf79384/src/dom_components/index.ts#L419) but its mechanism is quite simple so from what I see your approach to extend traits looks legit.

**Reply by jcenlo** [reply]
@artf Thank you very much for all the help WikiGrapesjs, without you this would be very difficult

I am a designer/frontend(layout) and i am starting to program more seriously (react,js...), and without all the help that you offer so actively, this would be much more difficult for me

I am doing my first project and this editor has a thousand possibilities, 
I am sure that you are going to help many people to learn, create businesses, etc..., offer them a better life, motivation,...

**Reply by artf** [reply]
Thanks @jcencina I'm glad to be helpful 😊
